substrate/tendermint/client/src/authority/gossip.rs

```diff
@@ -0,0 +1,67 @@
+use std::sync::{Arc, RwLock};
+
+use sp_core::Decode;
+use sp_runtime::traits::{Hash, Header, Block};
+
+use sc_network::PeerId;
+use sc_network_gossip::{Validator, ValidatorContext, ValidationResult};
+
+use tendermint_machine::{ext::SignatureScheme, SignedMessage};
+
+use crate::{TendermintValidator, validators::TendermintValidators};
+
+#[derive(Clone)]
+pub(crate) struct TendermintGossip<T: TendermintValidator> {
+  number: Arc<RwLock<u64>>,
+  signature_scheme: TendermintValidators<T>,
+}
+
+impl<T: TendermintValidator> TendermintGossip<T> {
+  pub(crate) fn new(number: Arc<RwLock<u64>>, signature_scheme: TendermintValidators<T>) -> Self {
+    TendermintGossip { number, signature_scheme }
+  }
+
+  pub(crate) fn topic(number: u64) -> <T::Block as Block>::Hash {
+    <<<T::Block as Block>::Header as Header>::Hashing as Hash>::hash(
+      &[b"Tendermint Block Topic".as_ref(), &number.to_le_bytes()].concat(),
+    )
+  }
+}
+
+impl<T: TendermintValidator> Validator<T::Block> for TendermintGossip<T> {
+  fn validate(
+    &self,
+    _: &mut dyn ValidatorContext<T::Block>,
+    _: &PeerId,
+    data: &[u8],
+  ) -> ValidationResult<<T::Block as Block>::Hash> {
+    let msg = match SignedMessage::<
+      u16,
+      T::Block,
+      <TendermintValidators<T> as SignatureScheme>::Signature,
+    >::decode(&mut &*data)
+    {
+      Ok(msg) => msg,
+      Err(_) => return ValidationResult::Discard,
+    };
+
+    if msg.number().0 < *self.number.read().unwrap() {
+      return ValidationResult::Discard;
+    }
+
+    // Verify the signature here so we don't carry invalid messages in our gossip layer
+    // This will cause double verification of the signature, yet that's a minimal cost
+    if !msg.verify_signature(&self.signature_scheme) {
+      return ValidationResult::Discard;
+    }
+
+    ValidationResult::ProcessAndKeep(Self::topic(msg.number().0))
```

Comment:
... this is guarded explicitly by the if statement. It's just not nested into it. Not only is that a non-standard pattern for this ecosystem, and one frequently advised against in general, your immediately suggested code isn't possible.

```rs
if msg.verify_signature(&self.signature_scheme) && !msg.number().0 < *self.number.read().unwrap() {
    ValidationResult::ProcessAndKeep(Self::topic(msg.number().0))
}
```
isn't valid as a return as it's incomplete. It'd need to be
```rs
if msg.verify_signature(&self.signature_scheme) && (!msg.number().0 < *self.number.read().unwrap()) {
    ValidationResult::ProcessAndKeep(Self::topic(msg.number().0))
} else {
  ValidationResult::Discard
}
```

This adds more lines, while not easily allowing expanding the conditional. Instead of just inserting a new if, we have to further extend that long conditional/do multiple layers of nesting inside of it. You also suggested combining two separate if statements, which alone isn't an issue, EXCEPT if those conditionals SHOULD have different return values, they can't be nested as such. While here, they both trigger discard, that's not always the case, and why the early return pattern remains the most flexible.